# <font color="#6e8ebc">**Data Sets**</font>
---

# <font color="#6e8ebc">**Índice**</font>

1. <font color="#6e8ebc">**Configuraciones**</font>
   * Importación de librerías
   * Paths
   * Paletas de colores
2. <font color="#6e8ebc">**Exploración inicial**</font>
   * Dataset Churn Bank
   * Dataset PaySim
3. <font color="#6e8ebc">**Manipulación de datos**</font>
   * Churn
     - Fechas
     - Clusters Iniciales
     - Clusters K-Means
     - Interpretación Clusters K-Means
4. <font color="#6e8ebc">**Transactions - Generación de datos**</font>
   * Chequeos generales
   * Distribuciones numéricas
   * Churn
   * Correlación
5. <font color="#6e8ebc">**App Interaction - Generación de datos**</font>
6. <font color="#6e8ebc">**Guardado Datasets**</font>
7. <font color="#6e8ebc">**Datasets Split**</font>

# <font color="#6e8ebc">**Cofiguraciones**</font>

## <font color="#ff743d">**Importación de librerías**</font>

In [92]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import random
import shutil

import kagglehub

from tqdm import tqdm

import os
from pathlib import Path
from typing import Tuple, Callable, List

import warnings

In [93]:
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None)

RANDOM_SEED = 42

## <font color="#ff743d">**Paths**</font>

In [94]:
# Obtiene el path actual
PROJECT_PATH = os.getcwd()

src = Path('src')
DATA_PATH = PROJECT_PATH / src

splited = Path('monthly_data')
MONTHLY_DATA_PATH = DATA_PATH / splited

tx = Path('tx')
TX_DATA_PATH = MONTHLY_DATA_PATH / tx

app = Path('app')
APP_DATA_PATH = MONTHLY_DATA_PATH / app

img = Path('img')
IMG_PATH = PROJECT_PATH / img

models = Path('models')
MODELS_PATH = PROJECT_PATH / models

reports = Path('reports')
REPORTS_PATH = PROJECT_PATH / reports

dirs = [DATA_PATH, MONTHLY_DATA_PATH, TX_DATA_PATH, APP_DATA_PATH, IMG_PATH, MODELS_PATH, REPORTS_PATH]


for directory in dirs:
    if not directory.exists():
        directory.mkdir(parents=True)
        print(f'Directorio creado:\n{directory}\n')
    else:
        print(f'El directorio ya existe:\n{directory}\n')

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src\monthly_data

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src\monthly_data\tx

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src\monthly_data\app

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\img

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\models

El directorio ya existe:
C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\reports



## <font color="#ff743d">**Paleta de Colores del Proyecto**</font>

**Naranjas**

<span style="display:inline-block; width:20px; height:20px; background:#993c00; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#cc5a00; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ff7b00; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ff9c40; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ffbf80; margin-right:5px;"></span>

**Rojos**

<span style="display:inline-block; width:20px; height:20px; background:#7a1c00; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#b23200; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#e74903; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ff743d; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#ffa68c; margin-right:5px;"></span>

**Azules**

<span style="display:inline-block; width:20px; height:20px; background:#001a37; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#002b5c; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#003873; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#315c99; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#6e8ebc; margin-right:5px;"></span>

**Celestes**

<span style="display:inline-block; width:20px; height:20px; background:#004752; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#007789; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#00b5d0; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#40cee6; margin-right:5px;"></span>
<span style="display:inline-block; width:20px; height:20px; background:#8ae6f4; margin-right:5px;"></span>

In [95]:
NARANJA_1, NARANJA_2, NARANJA_3, NARANJA_4, NARANJA_5 = '#993c00', '#cc5a00', '#ff7b00', '#ff9c40', '#ffbf80'
ROJO_1, ROJO_2, ROJO_3, ROJO_4, ROJO_5 = '#7a1c00', '#b23200', '#e74903', '#ff743d', '#ffa68c'
AZUL_1, AZUL_2, AZUL_3, AZUL_4, AZUL_5 = '#001a37', '#002b5c', '#003873', '#315c99', '#6e8ebc'
CIELO_1, CIELO_2, CIELO_3, CIELO_4, CIELO_5 = '#004752', '#007789', '#00b5d0', '#40cee6', '#8ae6f4'

# <font color="#6e8ebc">**Exploración inicial**</font>

## <font color="#ff743d">**Dataset Churn Bank**</font>

In [96]:
def obtener_datos_kagglehub(destination_path):
    """
    Descarga el dataset usando kagglehub y mueve el CSV a la carpeta del proyecto.
    """
    dataset_slug = "mathchi/churn-for-bank-customers"
    
    print(f"⬇️ Solicitando dataset: {dataset_slug} via kagglehub...")
    
    try:
        # 1. Descarga (kagglehub gestiona el caché automáticamente)
        path_cache = kagglehub.dataset_download(dataset_slug)
        print(f"✅ Descargado en caché: {path_cache}")
        
        # 2. Mover el archivo CSV a nuestra carpeta de proyecto (DATA_PATH)
        # Buscamos archivos CSV en la carpeta descargada
        archivos_csv = [f for f in os.listdir(path_cache) if f.endswith('.csv')]
        
        if not archivos_csv:
            print("⚠️ No se encontró ningún archivo CSV en el dataset descargado.")
            return

        for archivo in archivos_csv:
            origen = os.path.join(path_cache, archivo)
            destino = os.path.join(destination_path, archivo)
            
            # Copiamos/Sobreescribimos para asegurar que tenemos la última versión
            shutil.copy(origen, destino)
            print(f"📂 Archivo movido a: {destino}")

    except Exception as e:
        print(f"❌ Error al descargar con kagglehub: {e}")

# --- EJECUCIÓN ---
# Asegurarse de tener instalado: pip install kagglehub
obtener_datos_kagglehub(DATA_PATH)

try:
    csv_path = os.path.join(DATA_PATH, "churn.csv") 
    if os.path.exists(csv_path):
        df_churn = pd.read_csv(csv_path)
        print(f"\n✅ DataFrame cargado con éxito. Shape: {df_churn.shape}")
        print(df_churn.head(3))
except Exception as e:
    print("\n⚠️ El archivo se descargó, pero verifica el nombre exacto en la carpeta data/")

⬇️ Solicitando dataset: mathchi/churn-for-bank-customers via kagglehub...
✅ Descargado en caché: C:\Users\Ignacio\.cache\kagglehub\datasets\mathchi\churn-for-bank-customers\versions\1
📂 Archivo movido a: C:\Users\Ignacio\OneDrive\Documents\git-and-github\hackathon_ONE\src\churn.csv

✅ DataFrame cargado con éxito. Shape: (10000, 14)
   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.5

In [97]:
df_churn.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [6]:
df_churn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [7]:
df_churn.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

In [8]:
df_churn.describe()

,RowNumber,CustomerId,CreditScore,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
count,10000.00000,1.000000e+04,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000
mean,5000.50000,1.569094e+07,650.528800,38.921800,5.012800,76485.889288,1.530200,0.70550,0.515100,100090.239881,0.203700
std,2886.89568,7.193619e+04,96.653299,10.487806,2.892174,62397.405202,0.581654,0.45584,0.499797,57510.492818,0.402769
min,1.00000,1.556570e+07,350.000000,18.000000,0.000000,0.000000,1.000000,0.00000,0.000000,11.580000,0.000000
25%,2500.75000,1.562853e+07,584.000000,32.000000,3.000000,0.000000,1.000000,0.00000,0.000000,51002.110000,0.000000
50%,5000.50000,1.569074e+07,652.000000,37.000000,5.000000,97198.540000,1.000000,1.00000,1.000000,100193.915000,0.000000
75%,7500.25000,1.575323e+07,718.000000,44.000000,7.000000,127644.240000,2.000000,1.00000,1.000000,149388.247500,0.000000
max,10000.00000,1.581569e+07,850.000000,92.000000,10.000000,250898.090000,4.000000,1.00000,1.000000,199992.480000,1.000000


In [9]:
df_churn['Exited'].sum()

2037

In [10]:
df_churn['Exited'].value_counts(normalize=True)

Exited
0    0.7963
1    0.2037
Name: proportion, dtype: float64

In [11]:
df_churn['CustomerId'].nunique()

10000

In [12]:
path_paysim = kagglehub.dataset_download('ealaxi/paysim1')

print('Path to dataset PaySim:', path_paysim)

Path to dataset PaySim: C:\Users\Ignacio\.cache\kagglehub\datasets\ealaxi\paysim1\versions\2


In [13]:
paysim_file = Path('PS_20174392719_1491204439457_log.csv')
df_paysim = pd.read_csv(path_paysim / paysim_file)
df_paysim.head()

,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
0,1,PAYMENT,9839.64,C1231006815,170136.0,160296.36,M1979787155,0.0,0.0,0,0
1,1,PAYMENT,1864.28,C1666544295,21249.0,19384.72,M2044282225,0.0,0.0,0,0
2,1,TRANSFER,181.00,C1305486145,181.0,0.00,C553264065,0.0,0.0,1,0
3,1,CASH_OUT,181.00,C840083671,181.0,0.00,C38997010,21182.0,0.0,1,0
4,1,PAYMENT,11668.14,C2048537720,41554.0,29885.86,M1230701703,0.0,0.0,0,0


In [14]:
df_paysim.shape

(6362620, 11)

In [15]:
df_paysim.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


In [16]:
df_paysim.isnull().sum()

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
isFlaggedFraud    0
dtype: int64

In [17]:
df_paysim.describe()

,step,amount,oldbalanceOrg,newbalanceOrig,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
count,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06
mean,2.433972e+02,1.798619e+05,8.338831e+05,8.551137e+05,1.100702e+06,1.224996e+06,1.290820e-03,2.514687e-06
std,1.423320e+02,6.038582e+05,2.888243e+06,2.924049e+06,3.399180e+06,3.674129e+06,3.590480e-02,1.585775e-03
min,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,1.560000e+02,1.338957e+04,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
50%,2.390000e+02,7.487194e+04,1.420800e+04,0.000000e+00,1.327057e+05,2.146614e+05,0.000000e+00,0.000000e+00
75%,3.350000e+02,2.087215e+05,1.073152e+05,1.442584e+05,9.430367e+05,1.111909e+06,0.000000e+00,0.000000e+00
max,7.430000e+02,9.244552e+07,5.958504e+07,4.958504e+07,3.560159e+08,3.561793e+08,1.000000e+00,1.000000e+00


In [18]:
df_paysim['type'].unique()

array(['PAYMENT', 'TRANSFER', 'CASH_OUT', 'DEBIT', 'CASH_IN'],
      dtype=object)

In [19]:
df_paysim['type'].value_counts(normalize=True)

type
CASH_OUT    0.351663
PAYMENT     0.338146
CASH_IN     0.219923
TRANSFER    0.083756
DEBIT       0.006512
Name: proportion, dtype: float64

In [20]:
df_paysim['amount'].max()

92445516.64

In [21]:
df_churn['EstimatedSalary'].max()

199992.48

In [22]:
len(df_paysim) / df_churn['CustomerId'].nunique()

636.262

### <font color='#c37171'>**DataFrame Churn**</font>

* **Clientes unicos:** 10 000
  
* **Clientes Churn = Yes:** 2037 *(20% aprox)*

* **Max balance**: 250 898.09
* **Max estimated salary**: 199 992.48

### **Problemas iniciales**

El primer problema surge por supuesto porque los Ids no son los mismos, ya que no corresponden al mismo dataset. Uno cuenta con 10000 ids unicos (df_churn), el otro cuenta con 6353307 ids unicos (casi la misma cantidad de transacciones).

Otro problema que surge es que los montos son muy distintos entre ambos datasets. Se puede ver en el monto maximo de transaccion de PaySim y el balance maximo de Churn Otra cosa es que el step es de una hora, cuenta con 743 (correspondiente a un mes) cuando el tenure en el df_churn está medido en años, esto habrá que modificarlo de alguna manera.
<br>
### **Ideas iniciales**
<br>

> **TRANSACCIONES**

* Quedarme con todas las transacciones. Pero como el objetivo del problema es churn, no me interesan varias columnas. 

Consevaría:

**step:** lo transformaría en fechas en un rango quizás de 2 años ya que hay 743 steps<br>
**type:** tipo de transacciones<br>
**amount:** necesita ser reescalado, y las transacciones no pueden ser nunca mayores al saldo de la cuenta<br>
**nameOrig:** reemplazar por IDs de df_churn<br>

Estas dos no estoy seguro todavía:

**oldbalanceOrg**<br>
**newbalanceOrig**

<br>

> **IDs**

* Utilizar los IDs de **df_churn**, y buscar la manera de asignarlos de manera aleatoria a las transacciones. Esto deberá seguir cierta lógica ya que los clientes que se fueron no deberían tener transacciones en los últimos X meses.
Esto deberá respetar una logica donde se agrupen los clientes que tienen `Exited = 1` y que vaya variando, debe haber clientes con distintas fechas de abandono, entre 6 y 0 meses.

# <font color="#6e8ebc">**Manipulación de datos**</font>

## 🔑 <font color="#ff743d">**Principios para evitar “dataset basura”**</font>
---


### 1. Definir hipótesis de negocio realistas

Antes de tocar los datos, escribir supuestos que tengan sentido en un banco real:

- Clientes con menos interacciones en la app = más probabilidad de churn.
- Disminución en frecuencia de transacciones en los últimos meses = mayor riesgo de churn.
- Edad y salario influyen en la estabilidad del cliente (jóvenes y bajos salarios = más churn).
- Factores externos como inflación alta → más churn en clientes de bajo ingreso.

> 👉 Esto asegura que los patrones que generes no sean aleatorios, sino consistentes con lo que podría pasar en la realidad.

---

### 2. Usar correlaciones en lugar de ruido puro
En lugar de asignar variables externas o de interacción al azar:

- Introducir variables que estén correlacionadas con churn de forma controlada.

**Ejemplo:**
- Clientes que hicieron churn tienen un 40% menos de interacciones con la app en los últimos 3 meses.
- Clientes de churn tienen un 20% más de transacciones fallidas o rechazadas.
- Factores externos (ej: inflación simulada) afectan más a clientes con Balance < X.

> 👉 Esto genera señales diferenciales para el modelo.

---

### 3. Introducir variabilidad dentro de cada grupo

No todos los clientes que hicieron churn deben tener baja actividad. Introducir distribuciones:

- 60% de churners tienen muy baja interacción.
- 30% tienen actividad media.
- 10% siguen siendo activos hasta el final.

> 👉 Esto evita que el modelo aprenda reglas triviales (del tipo: “si transacciones = 0 → churn”).

---

### 4. Usar ruido controlado (stochastic noise)
Un poco de azar es necesario para que no todo sea determinístico, pero debe ser pequeño en comparación con la señal principal.

**Ejemplo:**  
Número de interacciones promedio de un churner = 20, con desviación estándar de ±5.

---

### 5. Validación manual antes del modelo
Antes de entrenar, explorar el dataset final con gráficos:

- Histograma de transacciones churn vs no churn.
- Interacciones app churn vs no churn.
- Factores externos según balance.

> 👉 Si se ven diferencias razonables entre grupos, entonces el dataset tiene patrones útiles.

---


## 🔧 <font color="#ff743d">**Overview de los 3 datasets**</font>

### 1. Churn (base)
- Target + atributos estáticos.

### 2. Transacciones
- Churners:
    * Clientes que muestran una caída de actividad en los últimos meses.
    * Clientes que abandonan de manera repentina
    * Clientes que a pesar de mostrar "actitud NO churner" abandonan
- No churners:
    * Clientes que no hicieron churn mantienen o aumentan su actividad.
    * Pausas (Ej.: vacaciones, simplemente no tuvieron necesidad, etc.)

### 3. Interacciones con la app
- Churners:
   * Clientes con menor login, menor tiempo en app, menos clicks en productos financieros.
- No churners:
    * Clientes con mayor consistencia en el uso.

## <font color="#ff743d">**Churn**</font>

In [23]:
df_churn_man = df_churn.copy()

### <font color="#ffbf80">**Fechas**</font>

In [24]:
periods = df_paysim['step'].nunique()
periods

743

In [25]:
end_date = pd.to_datetime('2025-09-30')
end_date

Timestamp('2025-09-30 00:00:00')

In [26]:
start_date = end_date - pd.Timedelta(days=periods-1)
start_date

Timestamp('2023-09-19 00:00:00')

In [27]:
# Rango de fechas de acuerdo a la cantida de steps dentro del df_paysim
fechas = pd.date_range(start=start_date, periods=periods, freq="D")
fechas

DatetimeIndex(['2023-09-19', '2023-09-20', '2023-09-21', '2023-09-22',
               '2023-09-23', '2023-09-24', '2023-09-25', '2023-09-26',
               '2023-09-27', '2023-09-28',
               ...
               '2025-09-21', '2025-09-22', '2025-09-23', '2025-09-24',
               '2025-09-25', '2025-09-26', '2025-09-27', '2025-09-28',
               '2025-09-29', '2025-09-30'],
              dtype='datetime64[ns]', length=743, freq='D')

In [28]:
enum = [i for i in range(1, 744)]

In [29]:
df_fechas = pd.DataFrame({'step': enum, 'fecha': fechas})
df_fechas

,step,fecha
0,1,2023-09-19
1,2,2023-09-20
2,3,2023-09-21
3,4,2023-09-22
4,5,2023-09-23
...,...,...
738,739,2025-09-26
739,740,2025-09-27
740,741,2025-09-28
741,742,2025-09-29


In [30]:
df_fechas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 743 entries, 0 to 742
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   step    743 non-null    int64         
 1   fecha   743 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(1)
memory usage: 11.7 KB


In [31]:
fechas_abandono = fechas[int((len(fechas) * 0.84)):]
fechas_abandono

DatetimeIndex(['2025-06-04', '2025-06-05', '2025-06-06', '2025-06-07',
               '2025-06-08', '2025-06-09', '2025-06-10', '2025-06-11',
               '2025-06-12', '2025-06-13',
               ...
               '2025-09-21', '2025-09-22', '2025-09-23', '2025-09-24',
               '2025-09-25', '2025-09-26', '2025-09-27', '2025-09-28',
               '2025-09-29', '2025-09-30'],
              dtype='datetime64[ns]', length=119, freq='D')

In [32]:
import random

np.random.seed(RANDOM_SEED)

abandono_cust = []

for exited in df_churn_man['Exited']:
    if exited == 1:
        if random.random() < 0.95:
            # 95% de churners en los últimos 4 meses
            rand_idx = int(np.random.triangular(0, len(fechas_abandono)-1, len(fechas_abandono)-1))
            abandono_cust.append(fechas_abandono[rand_idx])
        else:
            # 5 % churners en cualquier fecha del período
            abandono_cust.append(random.choice(fechas))
    else:
        abandono_cust.append(np.nan)

In [33]:
df_churn_man['ExitDate'] = abandono_cust

In [34]:
df_churn_man

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-08-15
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-09-27
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,9996,15606229,Obijiaku,771,France,Male,39,5,0.00,2,1,0,96270.64,0,NaT
9996,9997,15569892,Johnstone,516,France,Male,35,10,57369.61,1,1,1,101699.77,0,NaT
9997,9998,15584532,Liu,709,France,Female,36,7,0.00,1,0,1,42085.58,1,2025-07-04
9998,9999,15682355,Sabbatini,772,Germany,Male,42,3,75075.31,2,1,0,92888.52,1,2025-09-01


### <font color="#ffbf80">**Clusteres iniciales**</font>

Se asignarán scores de acuerdo a distintos aspectos del cliente:

* Tiempo: Edad y Antigüedad
* Riqueza: Balance, Salario Estimado y Puntaje Crediticio
* Compromiso: Número de productos, si Cuenta con Tarjeta de Crédito y si es Miembro Activo

Esto permitirá asignar un puntaje a cada cliente referente a cada aspecto, esto se utilizará luego para realizar una clusterización, y usar los clusters para asignar posteriormente las transacciones e interacciones con la app.

In [35]:
df_clusters = df_churn_man.copy()

In [36]:
df_clusters['cluster_time'] = pd.qcut(df_clusters['Tenure']*0.6 + df_clusters['Age']*0.4, q=3, labels=['Low','Medium','High'])

In [37]:
df_clusters['wealth_score'] = df_clusters['Balance']*0.5 + df_clusters['EstimatedSalary']*0.4 + df_clusters['CreditScore']*0.1

df_clusters['cluster_wealth'] = pd.qcut(df_clusters['wealth_score'], q=3, labels=['Low','Medium','High'])

In [38]:
df_clusters['engagement_score'] = df_clusters['NumOfProducts']*0.5 + df_clusters['HasCrCard']*0.2 + df_clusters['IsActiveMember']*0.3
df_clusters['cluster_engagement'] = pd.qcut(df_clusters['engagement_score'], q=3, labels=['Low','Medium','High'])

In [39]:
df_clusters

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-08-15,Medium,40601.452,Low,1.0,Low
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT,Medium,86981.762,Medium,0.8,Low
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-09-27,High,125453.228,High,1.7,High
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT,Low,37600.552,Low,1.0,Low
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT,Medium,94474.050,Medium,1.0,Low
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,9996,15606229,Obijiaku,771,France,Male,39,5,0.00,2,1,0,96270.64,0,NaT,Medium,38585.356,Low,1.2,Medium
9996,9997,15569892,Johnstone,516,France,Male,35,10,57369.61,1,1,1,101699.77,0,NaT,High,69416.313,Medium,1.0,Low
9997,9998,15584532,Liu,709,France,Female,36,7,0.00,1,0,1,42085.58,1,2025-07-04,Medium,16905.132,Low,0.8,Low
9998,9999,15682355,Sabbatini,772,Germany,Male,42,3,75075.31,2,1,0,92888.52,1,2025-09-01,Medium,74770.263,Medium,1.2,Medium


In [40]:
# Normalizar los scores para que estén en la misma escala
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

df_clusters[['wealth_score_norm', 'engagement_score_norm']] = scaler.fit_transform(df_clusters[['wealth_score', 'engagement_score']])

time_mapping = {'Low': 0, 'Medium': 0.5, 'High': 1}
df_clusters['cluster_time_num'] = df_clusters['cluster_time'].map(time_mapping).astype(float)

df_clusters.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement,wealth_score_norm,engagement_score_norm,cluster_time_num
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-08-15,Medium,40601.452,Low,1.0,Low,0.220009,0.25,0.5
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT,Medium,86981.762,Medium,0.8,Low,0.471943,0.15,0.5
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-09-27,High,125453.228,High,1.7,High,0.680918,0.60,1.0
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT,Low,37600.552,Low,1.0,Low,0.203708,0.25,0.0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT,Medium,94474.050,Medium,1.0,Low,0.512641,0.25,0.5


In [41]:
df_clusters['final_score'] = 0.4 * df_clusters['wealth_score_norm'] + 0.4 * df_clusters['engagement_score_norm'] + 0.2 * df_clusters['cluster_time_num']

### <font color="#ffbf80">**Clusteres K-Means**</font>

In [42]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=5, random_state=42)
df_clusters['cluster_final'] = kmeans.fit_predict(df_clusters[['final_score']])

In [43]:
df_clusters

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement,wealth_score_norm,engagement_score_norm,cluster_time_num,final_score,cluster_final
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-08-15,Medium,40601.452,Low,1.0,Low,0.220009,0.25,0.5,0.288004,2
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT,Medium,86981.762,Medium,0.8,Low,0.471943,0.15,0.5,0.348777,4
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-09-27,High,125453.228,High,1.7,High,0.680918,0.60,1.0,0.712367,3
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT,Low,37600.552,Low,1.0,Low,0.203708,0.25,0.0,0.181483,1
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT,Medium,94474.050,Medium,1.0,Low,0.512641,0.25,0.5,0.405056,4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,9996,15606229,Obijiaku,771,France,Male,39,5,0.00,2,1,0,96270.64,0,NaT,Medium,38585.356,Low,1.2,Medium,0.209058,0.35,0.5,0.323623,2
9996,9997,15569892,Johnstone,516,France,Male,35,10,57369.61,1,1,1,101699.77,0,NaT,High,69416.313,Medium,1.0,Low,0.376529,0.25,1.0,0.450612,0
9997,9998,15584532,Liu,709,France,Female,36,7,0.00,1,0,1,42085.58,1,2025-07-04,Medium,16905.132,Low,0.8,Low,0.091292,0.15,0.5,0.196517,1
9998,9999,15682355,Sabbatini,772,Germany,Male,42,3,75075.31,2,1,0,92888.52,1,2025-09-01,Medium,74770.263,Medium,1.2,Medium,0.405611,0.35,0.5,0.402245,4


### <font color="#ffbf80">**Interpretación clusters K-Means**</font>

In [44]:
clusters_count = df_clusters['cluster_final'].value_counts()
clusters_count

cluster_final
4    2874
2    2646
0    2230
1    1434
3     816
Name: count, dtype: int64

In [45]:
df_clusters_count = pd.DataFrame({'index':clusters_count.index, 'count': clusters_count})

In [46]:
df_clusters_count

,index,count
cluster_final,,
4,4,2874
2,2,2646
0,0,2230
1,1,1434
3,3,816


In [47]:
centroids_clusters_kmeans = kmeans.cluster_centers_
df_centroides = pd.DataFrame(centroids_clusters_kmeans, columns=['final_score'])
df_centroides.reset_index(inplace=True)

In [48]:
df_centroides

,index,final_score
0,0,0.494466
1,1,0.191013
2,2,0.299533
3,3,0.624389
4,4,0.395506


In [49]:
df_centroides = df_centroides.merge(df_clusters_count, on='index')
df_centroides = df_centroides.rename(columns={'index': 'cluster'})

In [50]:
df_centroides.sort_values('final_score')

,cluster,final_score,count
1,1,0.191013,1434
2,2,0.299533,2646
4,4,0.395506,2874
0,0,0.494466,2230
3,3,0.624389,816


In [51]:
def proporcion_labels(df):
    """
    Esta función tomará un df que debe corresponder a un unico cluster y
    retornará las proporciones de cada etiqueta de los clusters utilizados inicialmente
    """
    
    cluster_labels = ['cluster_time', 'cluster_wealth', 'cluster_engagement']
    counts = {}

    for cluster in cluster_labels:
        counts[cluster] = df[cluster].value_counts(normalize=True)

    return pd.DataFrame(counts)

#### <font color='#FDD2A9'>**Cluster 0**</font>

In [52]:
cluster_0 = df_clusters[df_clusters['cluster_final'] == 0]
prop_cluster_0 = proporcion_labels(cluster_0)
prop_cluster_0

,cluster_time,cluster_wealth,cluster_engagement
Low,0.070852,0.123318,0.503139
Medium,0.311211,0.343049,0.134978
High,0.617937,0.533632,0.361883


#### <font color='#FDD2A9'>**Observaciones Cluster 0:**</font>

Los clientes del **Cluster 0** cuentan con las siguientes características:
* Cuenta con un **score general** de: **0.494466** (2do más alto en el ranking)
* **Cantidad de clientes** pertenencientes al cluster: **2230 clientes**
* *cluster_time:* 60% para clientes con un alto score de tiempo (edad y permanencia) y 30% para término medio.
* *cluster_wealth:* 50% concentrado en clientes con buena salud económica, 35% para término medio.
* *cluster_engagement:* 50% con poco engagement, 35% con alto engagement.

Este es un buen perfil de cliente, ya que son clientes antiguos y con buen poder adquisitivo

#### <font color='#FDD2A9'>**Cluster 1**</font>

In [53]:
cluster_1 = df_clusters[df_clusters['cluster_final'] == 1]
prop_cluster_1 = proporcion_labels(cluster_1)
prop_cluster_1

,cluster_time,cluster_wealth,cluster_engagement
Low,0.82636,0.734310,0.722455
Medium,0.16318,0.243375,0.156206
High,0.01046,0.022315,0.121339


#### <font color='#FDD2A9'>**Observaciones Cluster 1:**</font>

Los clientes del **Cluster 1** cuentan con las siguientes características:

* Cuenta con un **score general** de: **0.191013** (ranking más bajo)
* **Cantidad de clientes** pertenencientes al cluster: **1434 clientes**
* *cluster_time:* 82% para clientes con un bajo score de tiempo (edad y permanencia) y 16% para término medio.
* *cluster_wealth:* 73% concentrado en clientes con baja salud económica, 24% para término medio.
* *cluster_engagement:* 72% con poco engagement, 12% con engagement término medio y 12% con alto engagement.

Este es el perfi de cliente con menor valor para la empresa, son clientes nuevos, de bajo poder adquisitivo, y con poco compromiso

#### <font color='#FDD2A9'>**Cluster 2**</font>

In [54]:
cluster_2 = df_clusters[df_clusters['cluster_final'] == 2]
prop_cluster_2 = proporcion_labels(cluster_2)
prop_cluster_2

,cluster_time,cluster_wealth,cluster_engagement
Low,0.522298,0.432729,0.642857
Medium,0.387755,0.365835,0.167800
High,0.089947,0.201436,0.189342


#### <font color='#FDD2A9'>**Observaciones Cluster 2:**</font>

Los clientes del **Cluster 2** cuentan con las siguientes características:

* Cuenta con un **score general** de: **0.299533**
* **Cantidad de clientes** pertenencientes al cluster: **2646 clientes**
* *cluster_time:* 52% para clientes con un bajo score de tiempo (edad y permanencia) y 38% para término medio.
* *cluster_wealth:* 20% concentrado en clientes con buena salud económica, 36% para término medio, y 43% para baja riqueza.
* *cluster_engagement:* 64% con poco engagement, 16% con engagement término medio y 19% con alto engagement.

Este es el perfi de cliente tiene poco compromiso, pero con potencial para fidelizar, ya que la riqueza esta distribuida de forma relativamente pareja, y son clientes nuevos y/o jovenes.

#### <font color='#FDD2A9'>**Cluster 3**</font>

In [55]:
cluster_3 = df_clusters[df_clusters['cluster_final'] == 3]
prop_cluster_3 = proporcion_labels(cluster_3)
prop_cluster_3

,cluster_time,cluster_wealth,cluster_engagement
Low,0.009804,0.030637,0.215686
Medium,0.151961,0.197304,0.123775
High,0.838235,0.772059,0.660539


#### <font color='#FDD2A9'>**Observaciones Cluster 3:**</font>

Los clientes del **Cluster 3** cuentan con las siguientes características:
	
* Cuenta con un **score general** de: **0.624389** (ranking más alto)
* **Cantidad de clientes** pertenencientes al cluster: **816 clientes** (cluster más pequeño)
* *cluster_time:* 83% para clientes con un alto score de tiempo (edad y permanencia) y 15% para término medio.
* *cluster_wealth:* 77% concentrado en clientes con buena salud económica, 19% para término medio, y 3% para baja riqueza.
* *cluster_engagement:* 66% con alto engagement, 12% con engagement término medio y 21% con bajo engagement.

Este es el perfil de cliente **"VIP"**. Son clientes antiguos y con muy buena salud económica (posiblemente también sean los clientes más adultos). Y con buen nivel de compromiso.

#### <font color='#FDD2A9'>**Cluster 4**</font>

In [56]:
cluster_4 = df_clusters[df_clusters['cluster_final'] == 4]
prop_cluster_4 = proporcion_labels(cluster_4)
prop_cluster_4

,cluster_time,cluster_wealth,cluster_engagement
Low,0.218163,0.290536,0.585247
Medium,0.452679,0.379610,0.153445
High,0.329158,0.329854,0.261308


#### <font color='#FDD2A9'>**Observaciones Cluster 4:**</font>

Los clientes del **Cluster 4** cuentan con las siguientes características:
	
* Cuenta con un **score general** de: **0.395506**
* **Cantidad de clientes** pertenencientes al cluster: **2874 clientes** (cluster más grande)
* *cluster_time:* 32% para clientes con un alto score de tiempo (edad y permanencia) y 45% para término medio y 21% de bajo score de tiempo.
* *cluster_wealth:* 32% de clientes con buena salud económica, 37% para término medio, y 29% para baja riqueza.
* *cluster_engagement:* 26% con alto engagement, 15% con engagement término medio y 58% con bajo engagement.

Este es el perfil de cliente **"Standard"**. Son clientes ni muy antiguos ni muy nuevos, ni muy ricos ni muy pobres, predomina un compromiso bajo.

In [57]:
etiquetas = {0: 'Valioso - Bajo compromiso',
             1: 'Nuevo - Poco Valor',
             2: 'Cliente potencial',
             3: 'VIP',
             4: 'Standard'}

In [58]:
df_clusters['CustomerSegment'] = df_clusters['cluster_final'].map(etiquetas)

In [59]:
df_clusters.sample(20)

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement,wealth_score_norm,engagement_score_norm,cluster_time_num,final_score,cluster_final,CustomerSegment
9850,9851,15626012,Obidimkpa,459,France,Male,26,4,149879.66,1,0,0,50016.17,0,NaT,Low,94992.198,Medium,0.5,Low,0.515456,0.00,0.0,0.206182,1,Nuevo - Poco Valor
9374,9375,15632538,Watson,658,Spain,Female,32,5,145553.07,1,1,1,31484.76,0,NaT,Low,85436.239,Medium,1.0,Low,0.463548,0.25,0.0,0.285419,2,Cliente potencial
5179,5180,15620000,Chambers,760,Germany,Male,34,6,121303.77,2,1,1,59325.21,0,NaT,Medium,84457.969,Medium,1.5,High,0.458234,0.50,0.5,0.483294,0,Valioso - Bajo compromiso
7379,7380,15607746,Belstead,573,France,Female,36,1,0.00,1,1,1,56905.38,0,NaT,Low,22819.452,Low,1.0,Low,0.123418,0.25,0.0,0.149367,1,Nuevo - Poco Valor
2020,2021,15565779,Kent,627,Germany,Female,30,6,57809.32,1,1,0,188258.49,0,NaT,Low,104270.756,High,0.7,Low,0.565856,0.10,0.0,0.266342,2,Cliente potencial
5992,5993,15682955,Capon,758,France,Female,32,2,84378.90,1,1,1,75396.43,0,NaT,Low,72423.822,Medium,1.0,Low,0.392866,0.25,0.0,0.257146,2,Cliente potencial
9612,9613,15794297,McKay,776,France,Male,36,1,0.00,2,1,0,53477.76,0,NaT,Low,21468.704,Low,1.2,Medium,0.116081,0.35,0.0,0.186432,1,Nuevo - Poco Valor
5701,5702,15724154,Manna,625,Germany,Female,49,4,128504.76,1,1,0,126812.63,1,2025-07-08,High,115039.932,High,0.7,Low,0.624353,0.10,1.0,0.489741,0,Valioso - Bajo compromiso
7141,7142,15643056,McMillan,755,Germany,Female,38,1,82083.52,1,0,1,10333.78,0,NaT,Low,45250.772,Low,0.8,Low,0.245264,0.15,0.0,0.158105,1,Nuevo - Poco Valor
4077,4078,15659236,Iadanza,781,Spain,Male,33,3,0.00,2,1,1,42556.33,0,NaT,Low,17100.632,Low,1.5,High,0.092354,0.50,0.0,0.236942,1,Nuevo - Poco Valor


In [60]:
df_clients = df_clusters.copy()

In [61]:
df_clients.groupby('CustomerSegment')['CustomerSegment'].value_counts()

CustomerSegment
Cliente potencial            2646
Nuevo - Poco Valor           1434
Standard                     2874
VIP                           816
Valioso - Bajo compromiso    2230
Name: count, dtype: int64

In [62]:
df_clients['CustomerSegment'].value_counts(normalize=True)

CustomerSegment
Standard                     0.2874
Cliente potencial            0.2646
Valioso - Bajo compromiso    0.2230
Nuevo - Poco Valor           0.1434
VIP                          0.0816
Name: proportion, dtype: float64

In [63]:
churn_yes = df_clients[df_clients['Exited'] == 1]
churn_yes.groupby('CustomerSegment')['CustomerSegment'].value_counts()

CustomerSegment
Cliente potencial            380
Nuevo - Poco Valor           178
Standard                     590
VIP                          329
Valioso - Bajo compromiso    560
Name: count, dtype: int64

### <font color="#ffbf80">**Generación de transacciones**</font>

A continuación, utilizando las etiquetas de ambos datasets, se asignarán las transacciones de manera aleatoria, pero respetando ciertas proporciones que reflejen la realidad.

**Nota:**

Las siguients proporciones fueron asignadas a mano, buscando una distribución lógica entre los distintos tipos de clientes y transacciones y como debería ser el comportamiento de estos.

Además, se asignarán factores aleatorios para introducir ruido estadístico al conjunto de datos

#### **Las siguients tablas se utilizarán como patrones a seguir en la asignación de transacciones a los clientes**

**Guía**

| Etiqueta	                  |	Número de transacciones	 |
|-----------------------------|--------------------------|
| VIP	                      | Alta	                 |
| Valioso - Bajo compromiso   | Media	                 |
| Standard              	  | Media	                 |
| Clientes potenciales	      | Baja	                 |
| Nuevos - Poco Valor         | Baja	                 |

**Transacciones**

| Tran Label  | Proporción |
|-------------|------------|
| Low         | 0.33       |
| High        | 0.30       |
| Medium      | 0.27       |
| VIP         | 0.10       |

**Distribución Transacciones**

| Cliente Label \ Tran Label | Low           | Medium          | High         | Vip        | Total      |
|----------------------------|---------------|-----------------|--------------|------------|------------|
| Nuevos - Poco Valor        | 10%           | 3%              | 1%           | 0.5%       | 14%        |
| Clientes potenciales       | 8%            | 8%              | 3%           | 1%         | 20%        |
| Standard              	 | 6.5%          | 5%              | 7%           | 1%         | 19.5%      |
| Valioso - Bajo compromiso  | 7.5%          | 8%              | 4%           | 1%         | 20.5%      |
| VIP	                     | 1%            | 3%              | 15%          | 6.5%       | 26%        |
| Total                      | 33%           | 27%             | 30%          | 10%        | 100%       |

# <font color="#6e8ebc">**Transactions - Generación de datos**</font>

In [64]:
import numpy as np
import pandas as pd

# --- CONFIGURACIÓN Y SEMILLA ---
RANDOM_SEED = 42
RNG = np.random.default_rng(RANDOM_SEED)

# --- HIPERPARÁMETROS ---
TX_VOLUME_NOISE = 0.15   # Bajamos un poco el ruido de volumen también

# Tipos de transacción
TX_TYPES = ['PAYMENT', 'TRANSFER', 'CASH_OUT', 'DEBIT', 'CASH_IN']

# Multiplicadores (Referencia para escalar el monto según el tipo)
TX_TYPE_MULTIPLIERS = {
    'PAYMENT': 0.3,    
    'TRANSFER': 1.0,   
    'CASH_OUT': 0.5,   # Bajamos un poco retiros
    'DEBIT': 0.15,     # Débitos más pequeños (compras diarias)
    'CASH_IN': 1.0     
}

# --- PARÁMETROS POR CLUSTER (MODIFICADO PARA ESTABILIDAD) ---
# Eliminamos 'mu_amount' y 'sigma_amount'.
# Usamos 'amount_volatility': Cuánto varía el monto respecto al promedio (0.1 = 10% de variación)
tx_cluster_params = {
    "Nuevo - Poco Valor": {
        "base_tx_per_month": 3,
        "salary_ratio": 0.04,     # 4% del sueldo por tx
        "amount_volatility": 0.1, # Muy estable (varía +/- 10%)
        "probs_type": [0.40, 0.10, 0.10, 0.30, 0.10] 
    },
    "Cliente potencial": {
        "base_tx_per_month": 6,
        "salary_ratio": 0.06,  
        "amount_volatility": 0.15,
        "probs_type": [0.35, 0.20, 0.10, 0.25, 0.10]
    },
    "Standard": {
        "base_tx_per_month": 10,
        "salary_ratio": 0.05,  
        "amount_volatility": 0.2, # Un poco más variable
        "probs_type": [0.30, 0.30, 0.10, 0.20, 0.10]
    },
    "Valioso - Bajo compromiso": {
        "base_tx_per_month": 8,
        "salary_ratio": 0.08, 
        "amount_volatility": 0.25,
        "probs_type": [0.20, 0.40, 0.20, 0.10, 0.10]
    },
    "VIP": {
        "base_tx_per_month": 20,
        "salary_ratio": 0.12, 
        "amount_volatility": 0.3, # Los ricos varían más sus gastos
        "probs_type": [0.15, 0.50, 0.15, 0.05, 0.15]
    }
}

# --- PARÁMETROS DE CHURN ---
LAST_N_MONTHS_DECAY = 4
DECAY_FINAL_FRAC_TX = 0.10
PROB_CHURN_REPENTINO_TX = 0.20

# --- PREPARACIÓN DEL GRID ---
freq = "M"
months = pd.period_range(start=start_date, end=end_date, freq=freq).to_timestamp('M')

# Traer datos y asegurar salario
clients_tx = df_clients[['CustomerId', 'CustomerSegment', 'EstimatedSalary', 'Exited', 'ExitDate', 'Tenure']].copy()
clients_tx['Tenure'] = clients_tx['Tenure'].astype(float)
mean_salary = clients_tx['EstimatedSalary'].mean()
clients_tx['EstimatedSalary'] = clients_tx['EstimatedSalary'].fillna(mean_salary)

# Grid Cliente x Mes
cust_ids = clients_tx['CustomerId'].values
n_cust = len(cust_ids)
n_months = len(months)

grid_tx = pd.DataFrame({
    'CustomerId': np.repeat(cust_ids, n_months),
    'month': np.tile(months.to_numpy(), n_cust)
})
grid_tx = grid_tx.merge(clients_tx, on='CustomerId', how='left')

# Mapa de churners repentinos
churners_repentinos_tx = (RNG.random(len(clients_tx[clients_tx['Exited'] == 1])) < PROB_CHURN_REPENTINO_TX)
client_repentino_map_tx = clients_tx[clients_tx['Exited'] == 1].set_index('CustomerId').assign(is_repentino=churners_repentinos_tx)['is_repentino'].to_dict()

# --- FUNCIÓN DE VOLUMEN (LAMBDA) ---
def compute_tx_lambda(row):
    params = tx_cluster_params.get(row['CustomerSegment'], tx_cluster_params['Standard'])
    base_lam = params['base_tx_per_month']
    
    # Factor Antigüedad (Suavizado)
    tenure_factor = (row['Tenure'] / 60.0) * 0.2 
    mean_lam = base_lam * (1 + tenure_factor)
    
    # Ruido Mensual (Reducido)
    noise = np.clip(RNG.normal(loc=1.0, scale=TX_VOLUME_NOISE), 0.7, 1.3)
    lam = max(0.1, mean_lam * noise)
    
    # Lógica de Churn (DECAY)
    if row['Exited'] == 1:
        exit_date = pd.to_datetime(row['ExitDate'])
        month_end = row['month'] + pd.offsets.MonthEnd(0)
        
        if month_end >= exit_date:
            return 0.0
            
        is_repentino = client_repentino_map_tx.get(row['CustomerId'], False)
        
        if not is_repentino:
            diff = (exit_date.to_period('M') - row['month'].to_period('M')).n
            if diff <= LAST_N_MONTHS_DECAY and diff > 0:
                frac = diff / float(LAST_N_MONTHS_DECAY)
                decay_factor = (DECAY_FINAL_FRAC_TX + (1 - DECAY_FINAL_FRAC_TX) * frac)
                lam = lam * decay_factor
                
    return lam

# --- GENERACIÓN DE TX ---
print("Calculando volumen...")
grid_tx['lambda_tx'] = grid_tx.apply(compute_tx_lambda, axis=1)
grid_tx['n_tx'] = RNG.poisson(grid_tx['lambda_tx'].values)
grid_tx = grid_tx[grid_tx['n_tx'] > 0].reset_index(drop=True)

print("Expandiendo...")
rep_counts = grid_tx['n_tx'].values
total_txs = rep_counts.sum()

tx_customer = np.repeat(grid_tx['CustomerId'].values, rep_counts)
tx_month = np.repeat(grid_tx['month'].values.astype('datetime64[ns]'), rep_counts)
tx_cluster_arr = np.repeat(grid_tx['CustomerSegment'].values, rep_counts)
tx_salary_arr = np.repeat(grid_tx['EstimatedSalary'].values, rep_counts)

# --- ASIGNACIÓN DE MONTOS ESTABLES ---
print("Asignando montos estables...")

tx_types = np.empty(total_txs, dtype=object)
tx_amounts = np.zeros(total_txs, dtype=float)

for clust_name, params in tx_cluster_params.items():
    mask = (tx_cluster_arr == clust_name)
    n_cluster = mask.sum()
    if n_cluster == 0: continue
    
    # 1. Tipos
    types_chosen = RNG.choice(TX_TYPES, size=n_cluster, p=params['probs_type'])
    tx_types[mask] = types_chosen
    
    # 2. Montos (Lógica Determinística con Ruido Normal)
    salaries = tx_salary_arr[mask]
    monthly_income = salaries / 12.0
    
    # Objetivo Base = IngresoMensual * RatioCluster
    base_target = monthly_income * params['salary_ratio']
    
    # Ajuste por Tipo de Tx
    type_factors = np.vectorize(TX_TYPE_MULTIPLIERS.get)(types_chosen)
    target_amounts = base_target * type_factors
    
    # Generar Ruido Normal (ej: media 1.0, desvío 0.1) -> Valores entre 0.8 y 1.2 mayormente
    # Esto garantiza que el monto esté SIEMPRE cerca del target del salario.
    volatility = params['amount_volatility']
    noise = RNG.normal(loc=1.0, scale=volatility, size=n_cluster)
    
    # Evitar negativos o cero extremos
    noise = np.clip(noise, 0.5, 2.0) 
    
    final_amounts = target_amounts * noise
    
    # Redondeo y Límites
    # Mínimo $5.00 para que parezca real, Máximo 1/3 del salario anual en una sola tx (seguridad)
    final_amounts = np.clip(final_amounts, 5.0, salaries / 3.0)
    tx_amounts[mask] = np.round(final_amounts, 2)

# Fechas
month_starts = pd.to_datetime(tx_month).to_period('M').to_timestamp()
month_ends = (pd.to_datetime(tx_month).to_period('M') + 1).to_timestamp() - pd.Timedelta(days=1)
days_in_month = (month_ends - month_starts).days + 1
rand_days = RNG.integers(0, days_in_month, size=total_txs)
tx_dates = month_starts + pd.to_timedelta(rand_days, unit='D')

tx_ids = np.arange(total_txs).astype(str)

df_transactions_synth = pd.DataFrame({
    'TransactionId': tx_ids,
    'CustomerId': tx_customer,
    'TransactionDate': tx_dates,
    'Amount': tx_amounts,
    'TransactionType': tx_types
})

df_transactions_synth = df_transactions_synth.sort_values(['CustomerId', 'TransactionDate']).reset_index(drop=True)

print("\n--- Resumen Transacciones Sintéticas (Estabilizado) ---")
print(df_transactions_synth['Amount'].describe().round(2))
print("\nPromedio por Tipo:")
print(df_transactions_synth.groupby('TransactionType')['Amount'].mean().round(2))

Calculando volumen...
Expandiendo...
Asignando montos estables...

--- Resumen Transacciones Sintéticas (Estabilizado) ---
count    2043646.00
mean         463.83
std          528.88
min            3.86
25%           97.51
50%          255.15
75%          646.78
max         3982.18
Name: Amount, dtype: float64

Promedio por Tipo:
TransactionType
CASH_IN     760.13
CASH_OUT    383.20
DEBIT        78.02
PAYMENT     176.15
TRANSFER    819.87
Name: Amount, dtype: float64


In [65]:
# 1. Calcular la fecha de la última transacción por cliente en df_transactions_synth
# La columna de fecha en df_transactions_synth es 'transaction_date'.
last_tx_date = df_transactions_synth.groupby('CustomerId')['TransactionDate'].max().reset_index(name='LastTransactionDate')

# 2. Filtrar clientes que abandonaron y unirlos con su última transacción
df_churned = df_clients[df_clients['Exited'] == 1].copy()
# Seleccionamos las columnas relevantes y realizamos el merge
df_churned_check = df_churned[['CustomerId', 'ExitDate', 'Exited']].merge(
    last_tx_date, on='CustomerId', how='left'
)

# 3. Realizar la comprobación de integridad
# El error ocurre si la ExitDate es ANTES de la LastTransactionDate.
df_churned_check['TemporalIntegrityError'] = df_churned_check['ExitDate'] < df_churned_check['LastTransactionDate']

# 4. Reportar los casos problemáticos (donde el error es True)
integrity_violations = df_churned_check[df_churned_check['TemporalIntegrityError'] == True]

print(f"Clientes que abandonaron ('Exited'=1): {df_churned.shape[0]}")
print(f"Clientes con violación de integridad temporal: {integrity_violations.shape[0]}")

if integrity_violations.empty:
    print("\n✅ ¡Validación completada! No se encontraron transacciones después de las fechas de salida registradas.")
else:
    print("\n🚨 ¡ATENCIÓN! Se encontraron violaciones de integridad. Primeros 5 casos:")
    # Usamos .dt.date para imprimir las fechas de forma limpia
    violations_clean = integrity_violations[['CustomerId', 'ExitDate', 'LastTransactionDate']].copy()
    violations_clean['ExitDate'] = violations_clean['ExitDate'].dt.date
    violations_clean['LastTransactionDate'] = violations_clean['LastTransactionDate'].dt.date
    
    print(violations_clean.head())
    print("\nEstas filas indican que el cliente tuvo actividad de transacción posterior a su fecha de salida ('ExitDate').")

Clientes que abandonaron ('Exited'=1): 2037
Clientes con violación de integridad temporal: 0

✅ ¡Validación completada! No se encontraron transacciones después de las fechas de salida registradas.


In [66]:
df_transactions_synth.sort_values(by='TransactionDate', inplace=True)

# <font color="#6e8ebc">**App Interaction - Generación de datos**</font>

In [67]:
import numpy as np
import pandas as pd
from datetime import datetime
import uuid

# --- REPRODUCIBILIDAD ---
# Es fundamental para que los resultados sean consistentes en cada ejecución.
RANDOM_SEED = 42
RNG = np.random.default_rng(RANDOM_SEED)

# --- CONFIGURACIÓN GLOBAL DE RUIDO ---
# Estos parámetros añaden variabilidad y realismo al comportamiento de los usuarios.
LAMBDA_NOISE_SIGMA = 0.20 # Variabilidad en la actividad mensual de cada cliente.
DECAY_NOISE_SIGMA = 0.25   # Variabilidad en el decaimiento de actividad de los churners.
PROB_NOISE_SIGMA = 0.15    # Variabilidad en las acciones que realiza un usuario en cada sesión.

# --- DEFINICIONES DE TIEMPO Y CLUSTERS ---
# (Asegúrate de que 'start_date' y 'end_date' estén definidas previamente)
freq = "M"  # Agruparemos la actividad por Meses.
months = pd.period_range(start=start_date, end=end_date, freq=freq).to_timestamp('M')

# nombres de clusters según tu df_clients
clusters = [
    "Nuevo - Poco Valor",
    "Cliente potencial",
    "Standard",
    "Valioso - Bajo compromiso",
    "VIP"
]

# parámetros base por cluster (ejemplo; ajustalos)
cluster_params = {
    "Nuevo - Poco Valor": {
        "base_sessions_per_month": 1.5,
        "duration_mu": 2.0,  # log-normal mu
        "duration_sigma": 0.6,
        "p_transfer": 0.02,
        "p_payment": 0.05,
        "p_invest": 0.01,
        "p_push_open": 0.10,
        "p_failed_login": 0.01,
    },
    "Cliente potencial": {
        "base_sessions_per_month": 3.0,
        "duration_mu": 2.3,
        "duration_sigma": 0.7,
        "p_transfer": 0.06,
        "p_payment": 0.08,
        "p_invest": 0.02,
        "p_push_open": 0.15,
        "p_failed_login": 0.02,
    },
    "Standard": {
        "base_sessions_per_month": 4.0,
        "duration_mu": 2.7,
        "duration_sigma": 0.8,
        "p_transfer": 0.10,
        "p_payment": 0.12,
        "p_invest": 0.03,
        "p_push_open": 0.18,
        "p_failed_login": 0.03,
    },
    "Valioso - Bajo compromiso": {
        "base_sessions_per_month": 3.5,
        "duration_mu": 2.6,
        "duration_sigma": 0.8,
        "p_transfer": 0.12,
        "p_payment": 0.10,
        "p_invest": 0.05,
        "p_push_open": 0.12,
        "p_failed_login": 0.02,
    },
    "VIP": {
        "base_sessions_per_month": 8.0,
        "duration_mu": 3.0,
        "duration_sigma": 0.9,
        "p_transfer": 0.25,
        "p_payment": 0.20,
        "p_invest": 0.15,
        "p_push_open": 0.35,
        "p_failed_login": 0.01,
    }
}


# --- PARÁMETROS DE COMPORTAMIENTO DE CHURN Y TENURE ---
prob_churn_repentino_app = 0.20 # 20% de los churners se irán "de golpe".
last_n_months_decay = 6       # Ventana de 6 meses donde la actividad de un churner decae.
decay_final_frac = 0.05       # La actividad cae hasta un 5% de lo normal justo antes de irse.
tenure_max = df_clients['Tenure'].max()
tenure_weight = 0.5           # Influencia de la antigüedad del cliente en su actividad.

# ==============================================================================
# === (NUEVO) HIPERPARÁMETROS DE PAUSA PARA NO CHURNERS ===
# ==============================================================================
PROB_PAUSA_NO_CHURNER_APP = 0.15  # 15% de los clientes fieles tendrán una pausa.
PAUSA_DURACION_MIN_MESES = 2       # Duración mínima de la pausa en meses.
PAUSA_DURACION_MAX_MESES = 4       # Duración máxima de la pausa en meses.
# ==============================================================================

# --- 1) PREPARAR BASE CLIENTE x MES ---
clients = df_clients[['CustomerId','CustomerSegment','Exited','ExitDate','Tenure']].copy()
clients['Tenure'] = clients['Tenure'].astype(float)

# Crear tabla con una fila por cada cliente y cada mes de la simulación.
cust_ids = clients['CustomerId'].values
n_cust = len(cust_ids)
n_months = len(months)
grid = pd.DataFrame({
    'CustomerId': np.repeat(cust_ids, n_months),
    'month': np.tile(months.to_numpy(), n_cust)
})
grid = grid.merge(clients, on='CustomerId', how='left')

# Mapa de churners repentinos (sin cambios).
churners_repentinos = (RNG.random(len(clients[clients['Exited'] == 1])) < prob_churn_repentino_app)
client_repentino_map = clients[clients['Exited'] == 1].set_index('CustomerId').assign(is_repentino=churners_repentinos)['is_repentino'].to_dict()


# ==============================================================================
# === (NUEVO) PRE-CÁLCULO DE LAS PAUSAS DE ACTIVIDAD PARA CLIENTES FIELES ===
# ==============================================================================
# 1. Seleccionar qué clientes no-churners tendrán una pausa.
no_churners = clients[clients['Exited'] == 0].copy()
clientes_con_pausa_ids = no_churners.sample(frac=PROB_PAUSA_NO_CHURNER_APP, random_state=RANDOM_SEED)['CustomerId']

# 2. Para cada cliente seleccionado, definir el inicio y fin de su pausa.
pause_info = []
total_meses = len(months)
for cid in clientes_con_pausa_ids:
    duracion_pausa = np.random.randint(PAUSA_DURACION_MIN_MESES, PAUSA_DURACION_MAX_MESES + 1)
    if total_meses > duracion_pausa:
        start_offset = np.random.randint(0, total_meses - duracion_pausa)
        pausa_inicio = months[start_offset]
        pausa_fin = months[start_offset + duracion_pausa - 1]
        pause_info.append({'CustomerId': cid, 'pause_start': pausa_inicio, 'pause_end': pausa_fin})

# 3. Unir la información de las pausas a la tabla principal `grid`.
if pause_info:
    df_pausas = pd.DataFrame(pause_info)
    grid = grid.merge(df_pausas, on='CustomerId', how='left')
else:
    grid['pause_start'] = pd.NaT
    grid['pause_end'] = pd.NaT

grid['pause_start'] = grid['pause_start'].fillna(pd.NaT)
grid['pause_end'] = grid['pause_end'].fillna(pd.NaT)
# ==============================================================================


# --- FUNCIÓN DE CÁLCULO DE ACTIVIDAD MENSUAL (MODIFICADA) ---
def compute_lambda_row(row):
    params = cluster_params.get(row['CustomerSegment'], cluster_params['Standard'])
    
    # Cálculo de Lambda Base con ruido (sin cambios).
    base_lam = params['base_sessions_per_month']
    tenure_factor = (row['Tenure'] / tenure_max) * tenure_weight
    mean_lam = base_lam * (1 + tenure_factor)
    noise_factor_base = np.clip(RNG.normal(loc=1.0, scale=LAMBDA_NOISE_SIGMA), 0.5, 2.0)
    lam = max(0.01, mean_lam * noise_factor_base)

    # ==========================================================================
    # === (MODIFICADO) LÓGICA PARA NO-CHURNERS CON PAUSAS ===
    # ==========================================================================
    if row['Exited'] == 0:
        is_in_pause = pd.notna(row['pause_start']) and (row['month'] >= row['pause_start']) and (row['month'] <= row['pause_end'])
        
        if is_in_pause:
            return 0.0  # Si está en pausa, la actividad de este mes es CERO.
        else:
            return lam  # Si no, tiene su actividad normal.
    # ==========================================================================
            
    # --- LÓGICA PARA CHURNERS (Sin cambios) ---
    else: # (row['Exited'] == 1)
        exit_date = pd.to_datetime(row['ExitDate'])
        month_end = row['month'] + pd.offsets.MonthEnd(0)
        
        if month_end >= exit_date:
            return 0.0
        
        is_repentino = client_repentino_map.get(row['CustomerId'], False)
        if is_repentino:
            return lam
        else:
            months_before = int(((exit_date.to_period('M') - row['month'].to_period('M')).n))
            if months_before <= last_n_months_decay and months_before > 0:
                frac = months_before / float(last_n_months_decay)
                base_decay_factor = (decay_final_frac + (1 - decay_final_frac) * frac)
                noise_decay_factor = np.clip(RNG.normal(loc=1.0, scale=DECAY_NOISE_SIGMA), 0.5, 1.5)
                lam = lam * base_decay_factor * noise_decay_factor
                return max(0.01, lam)
    
    return lam # Devuelve el lambda para los meses de churners fuera de la ventana de decaimiento


# --- APLICACIÓN DE LA FUNCIÓN Y GENERACIÓN DE DATOS (SIN CAMBIOS) ---

# Aplicar la función mejorada para calcular la actividad mensual de cada cliente.
grid['lambda'] = grid.apply(compute_lambda_row, axis=1)

# 3) Samplear el número de sesiones por cliente-mes usando una distribución de Poisson.
grid['n_sessions'] = RNG.poisson(grid['lambda'].values)

# Eliminar meses sin actividad para optimizar memoria.
grid = grid[grid['n_sessions'] > 0].reset_index(drop=True)

# 4) Expandir a nivel de sesión.
rep_counts = grid['n_sessions'].values
total_sessions = rep_counts.sum()
print(f"Total sesiones generadas: {total_sessions:,}")

session_customer = np.repeat(grid['CustomerId'].values, rep_counts)
session_month = np.repeat(grid['month'].values.astype('datetime64[ns]'), rep_counts)

# 5) Generar propiedades para cada sesión.
cust_to_cluster = clients.set_index('CustomerId')['CustomerSegment'].to_dict()
session_cluster = np.vectorize(cust_to_cluster.get)(session_customer)

# Duración de la sesión.
mu_map = {k: v['duration_mu'] for k, v in cluster_params.items()}
sigma_map = {k: v['duration_sigma'] for k, v in cluster_params.items()}
mu_arr = np.vectorize(mu_map.get)(session_cluster)
sigma_arr = np.vectorize(sigma_map.get)(session_cluster)
durations = np.clip(RNG.lognormal(mean=mu_arr, sigma=sigma_arr), 0.5, 180.0)

# Probabilidades base de cada evento.
p_transfer_base = np.vectorize(lambda cl: cluster_params[cl]['p_transfer'])(session_cluster)
p_payment_base  = np.vectorize(lambda cl: cluster_params[cl]['p_payment'])(session_cluster)
p_invest_base   = np.vectorize(lambda cl: cluster_params[cl]['p_invest'])(session_cluster)
p_push_base     = np.vectorize(lambda cl: cluster_params[cl]['p_push_open'])(session_cluster)
p_fail_base     = np.vectorize(lambda cl: cluster_params[cl]['p_failed_login'])(session_cluster)

# Inyectar ruido a las probabilidades.
noise_factors = np.clip(RNG.normal(loc=1.0, scale=PROB_NOISE_SIGMA, size=total_sessions), 0.5, 1.5)
p_transfer = np.clip(p_transfer_base * noise_factors, 0.001, 0.999)
p_payment  = np.clip(p_payment_base * noise_factors, 0.001, 0.999)
p_invest   = np.clip(p_invest_base * noise_factors, 0.001, 0.999)
p_push     = np.clip(p_push_base * noise_factors, 0.001, 0.999)
p_fail     = np.clip(p_fail_base * noise_factors, 0.001, 0.999)

# Muestreo de eventos para cada sesión.
used_transfer = RNG.random(total_sessions) < p_transfer
used_payment  = RNG.random(total_sessions) < p_payment
used_invest   = RNG.random(total_sessions) < p_invest
opened_push   = RNG.random(total_sessions) < p_push
failed_login  = RNG.random(total_sessions) < p_fail

# Asignar fechas aleatorias dentro de cada mes.
month_starts = pd.to_datetime(session_month).to_period('M').to_timestamp()
month_ends = (pd.to_datetime(session_month).to_period('M') + 1).to_timestamp() - pd.Timedelta(days=1)
days_in_month = (month_ends - month_starts).days + 1
rand_days = RNG.integers(0, days_in_month, size=total_sessions)
session_dates = month_starts + pd.to_timedelta(rand_days, unit='D')

# Crear IDs de sesión.
session_ids = np.arange(total_sessions).astype(str)

# Ensamblar el DataFrame final a nivel de sesión.
df_sessions_synth = pd.DataFrame({
    'SessionId': session_ids,
    'CustomerId': session_customer,
    'SessionDate': session_dates,
    'DurationMin': durations,
    'UsedTransfer': used_transfer.astype(int),
    'UsedPayment': used_payment.astype(int),
    'UsedInvest': used_invest.astype(int),
    'OpenedPush': opened_push.astype(int),
    'FailedLogin': failed_login.astype(int)
})

# 6) Agregados por cliente a nivel mensual.
df_sessions_synth['month'] = df_sessions_synth['SessionDate'].dt.to_period('M').dt.to_timestamp()


# 8) Validaciones y resumen final.
print("\n--- Resumen de Datos Generados ---")
print(f"Sesiones totales: {len(df_sessions_synth):,}")

Total sesiones generadas: 1,073,889

--- Resumen de Datos Generados ---
Sesiones totales: 1,073,889


In [68]:
# NOTA: Asegúrate de que df_sessions_synth y df_clients estén disponibles en este punto.

# 1. Calcular la fecha de la última sesión por cliente en df_sessions_synth
# La columna de fecha en df_sessions_synth es 'date'.
last_session_date = df_sessions_synth.groupby('CustomerId')['SessionDate'].max().reset_index(name='LastSessionDate')

# 2. Filtrar clientes que abandonaron y unirlos con su última sesión
df_churned = df_clients[df_clients['Exited'] == 1].copy()
# Seleccionamos las columnas relevantes y realizamos el merge
df_churned_check = df_churned[['CustomerId', 'ExitDate', 'Exited']].merge(
    last_session_date, on='CustomerId', how='left'
)

# 3. Realizar la comprobación de integridad
# El error ocurre si la ExitDate es ANTES de la LastSessionDate.
df_churned_check['TemporalIntegrityError'] = df_churned_check['ExitDate'] < df_churned_check['LastSessionDate']

# 4. Reportar los casos problemáticos (donde el error es True)
integrity_violations = df_churned_check[df_churned_check['TemporalIntegrityError'] == True]

print(f"Clientes que abandonaron ('Exited'=1): {df_churned.shape[0]}")
print(f"Clientes con violación de integridad temporal: {integrity_violations.shape[0]}")

if integrity_violations.empty:
    print("\n✅ ¡Validación completada! No se encontraron sesiones después de las fechas de salida registradas.")
else:
    print("\n🚨 ¡ATENCIÓN! Se encontraron violaciones de integridad. Primeros 5 casos:")
    # Usamos .dt.date para imprimir las fechas de forma limpia
    violations_clean = integrity_violations[['CustomerId', 'ExitDate', 'LastSessionDate']].copy()
    violations_clean['ExitDate'] = violations_clean['ExitDate'].dt.date
    violations_clean['LastSessionDate'] = violations_clean['LastSessionDate'].dt.date
    
    print(violations_clean.head())
    print("\nEstas filas indican que el cliente tuvo actividad de sesión posterior a su fecha de salida ('ExitDate').")

Clientes que abandonaron ('Exited'=1): 2037
Clientes con violación de integridad temporal: 0

✅ ¡Validación completada! No se encontraron sesiones después de las fechas de salida registradas.


In [69]:
df_sessions_synth.sort_values(by='SessionDate', inplace=True)

In [70]:
df_sessions_synth

,SessionId,CustomerId,SessionDate,DurationMin,UsedTransfer,UsedPayment,UsedInvest,OpenedPush,FailedLogin,month
249547,249547,15664639,2023-09-01,17.303132,0,1,0,0,0,2023-09-01
979540,979540,15567339,2023-09-01,18.057587,0,0,1,0,0,2023-09-01
130327,130327,15812272,2023-09-01,35.082076,0,0,0,0,0,2023-09-01
1068951,1068951,15655952,2023-09-01,59.944415,0,0,0,0,0,2023-09-01
979306,979306,15727391,2023-09-01,14.494765,1,0,0,0,0,2023-09-01
...,...,...,...,...,...,...,...,...,...,...
100673,100673,15790782,2025-09-30,6.061282,0,1,0,0,0,2025-09-01
647206,647206,15709917,2025-09-30,6.976754,0,0,0,0,0,2025-09-01
682268,682268,15593454,2025-09-30,3.311562,0,0,0,0,0,2025-09-01
647449,647449,15799494,2025-09-30,74.662854,1,1,0,0,0,2025-09-01


# <font color="#6e8ebc">**Salud del Dataset**</font>

## <font color="#ff743d">**Chequeos Generales**</font>

In [71]:
"""
# Clientes
print("Clientes - info general:")
print(df_clients_filtered.info())
print(df_clients_filtered.describe())

# Transacciones
print("\nTransacciones - info general:")
print(df_tx_scaled.info())
print(df_tx_scaled.describe())

# Chequear duplicados
print("\nDuplicados clientes:", df_clients_filtered.duplicated(subset=['CustomerId']).sum())
print("Duplicados transacciones:", df_tx_scaled.duplicated(subset=['CustomerId','date','amount']).sum())

# Chequear missing values
print("\nMissing en clientes:\n", df_clients_filtered.isna().sum())
print("\nMissing en transacciones:\n", df_tx_scaled.isna().sum())
"""

'\n# Clientes\nprint("Clientes - info general:")\nprint(df_clients_filtered.info())\nprint(df_clients_filtered.describe())\n\n# Transacciones\nprint("\nTransacciones - info general:")\nprint(df_tx_scaled.info())\nprint(df_tx_scaled.describe())\n\n# Chequear duplicados\nprint("\nDuplicados clientes:", df_clients_filtered.duplicated(subset=[\'CustomerId\']).sum())\nprint("Duplicados transacciones:", df_tx_scaled.duplicated(subset=[\'CustomerId\',\'date\',\'amount\']).sum())\n\n# Chequear missing values\nprint("\nMissing en clientes:\n", df_clients_filtered.isna().sum())\nprint("\nMissing en transacciones:\n", df_tx_scaled.isna().sum())\n'

## <font color="#ff743d">**Distribuciones Numéricas**</font>

In [72]:
"""
# Definir colores
main_color = BORDO_1
secondary = BORDO_2
kde_color = CREMA_1
"""

'\n# Definir colores\nmain_color = BORDO_1\nsecondary = BORDO_2\nkde_color = CREMA_1\n'

In [73]:
"""
numeric_cols_clients = ['CreditScore','Age','Tenure','Balance','NumOfProducts','HasCrCard','IsActiveMember','EstimatedSalary']
numeric_cols_tx = ['amount','amount_clipped','amount_ratio','amount_scaled']

# Histogramas clientes
plt.figure(figsize=(14,10))
for i,col in enumerate(numeric_cols_clients):
    plt.subplot(3,3,i+1)
    sns.histplot(df_clients_clean[col], bins=30, kde=True,
                 color=main_color,
                 line_kws={'color': kde_color, 'linewidth': 2})
    plt.title(col)
plt.tight_layout()
plt.savefig(os.path.join(IMG_PATH, "df_clients_numerical_dist.png"), dpi=300)
plt.show()
"""

'\nnumeric_cols_clients = [\'CreditScore\',\'Age\',\'Tenure\',\'Balance\',\'NumOfProducts\',\'HasCrCard\',\'IsActiveMember\',\'EstimatedSalary\']\nnumeric_cols_tx = [\'amount\',\'amount_clipped\',\'amount_ratio\',\'amount_scaled\']\n\n# Histogramas clientes\nplt.figure(figsize=(14,10))\nfor i,col in enumerate(numeric_cols_clients):\n    plt.subplot(3,3,i+1)\n    sns.histplot(df_clients_clean[col], bins=30, kde=True,\n                 color=main_color,\n                 line_kws={\'color\': kde_color, \'linewidth\': 2})\n    plt.title(col)\nplt.tight_layout()\nplt.savefig(os.path.join(IMG_PATH, "df_clients_numerical_dist.png"), dpi=300)\nplt.show()\n'

## <font color="#ff743d">**Churn**</font>

In [74]:
"""
# Pie chart distribución churn
churn_counts = df_clients_clean['Exited'].value_counts(normalize=True)

plt.figure(figsize=(6,6))
colors = [main_color, kde_color]  # Exited=0 → kde_color, Exited=1 → main_color
churn_counts.plot(kind='pie', autopct='%1.1f%%', startangle=90, colors=exited_colors,
                  labels=['No Churn (0)', 'Churn (1)'])
plt.ylabel("")  # quitar etiqueta del eje
plt.title("Distribución de Churn")
plt.savefig(os.path.join(IMG_PATH, "churn_distribution_pie.png"), dpi=300)
plt.show()
"""

'\n# Pie chart distribución churn\nchurn_counts = df_clients_clean[\'Exited\'].value_counts(normalize=True)\n\nplt.figure(figsize=(6,6))\ncolors = [main_color, kde_color]  # Exited=0 → kde_color, Exited=1 → main_color\nchurn_counts.plot(kind=\'pie\', autopct=\'%1.1f%%\', startangle=90, colors=exited_colors,\n                  labels=[\'No Churn (0)\', \'Churn (1)\'])\nplt.ylabel("")  # quitar etiqueta del eje\nplt.title("Distribución de Churn")\nplt.savefig(os.path.join(IMG_PATH, "churn_distribution_pie.png"), dpi=300)\nplt.show()\n'

## <font color="#ff743d">**Correlación**</font>

In [75]:
"""
corr_columns = numeric_cols_clients
corr_columns.append('Exited')
plt.figure(figsize=(10,8))
sns.heatmap(df_clients_clean[corr_columns].corr(), annot=True, fmt=".2f", cmap="coolwarm")
plt.title("Correlación variables clientes")
plt.savefig(os.path.join(IMG_PATH, "corr_matrix.png"), dpi=300)
plt.show()
"""

'\ncorr_columns = numeric_cols_clients\ncorr_columns.append(\'Exited\')\nplt.figure(figsize=(10,8))\nsns.heatmap(df_clients_clean[corr_columns].corr(), annot=True, fmt=".2f", cmap="coolwarm")\nplt.title("Correlación variables clientes")\nplt.savefig(os.path.join(IMG_PATH, "corr_matrix.png"), dpi=300)\nplt.show()\n'

# <font color="#6e8ebc">**Guardado Datasets**</font>

In [76]:
df_clients.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,ExitDate,cluster_time,wealth_score,cluster_wealth,engagement_score,cluster_engagement,wealth_score_norm,engagement_score_norm,cluster_time_num,final_score,cluster_final,CustomerSegment
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1,2025-08-15,Medium,40601.452,Low,1.0,Low,0.220009,0.25,0.5,0.288004,2,Cliente potencial
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0,NaT,Medium,86981.762,Medium,0.8,Low,0.471943,0.15,0.5,0.348777,4,Standard
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1,2025-09-27,High,125453.228,High,1.7,High,0.680918,0.60,1.0,0.712367,3,VIP
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0,NaT,Low,37600.552,Low,1.0,Low,0.203708,0.25,0.0,0.181483,1,Nuevo - Poco Valor
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0,NaT,Medium,94474.050,Medium,1.0,Low,0.512641,0.25,0.5,0.405056,4,Standard


In [77]:
df_clients.columns

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited', 'ExitDate',
       'cluster_time', 'wealth_score', 'cluster_wealth', 'engagement_score',
       'cluster_engagement', 'wealth_score_norm', 'engagement_score_norm',
       'cluster_time_num', 'final_score', 'cluster_final', 'CustomerSegment'],
      dtype='object')

In [78]:
df_clients = df_clients.drop(['cluster_time', 'wealth_score', 'cluster_wealth', 'engagement_score',
       'cluster_engagement', 'wealth_score_norm', 'engagement_score_norm',
       'cluster_time_num', 'final_score', 'cluster_final'], axis=1)

In [79]:
df_transactions_synth['TransactionDate'].min()

Timestamp('2023-09-01 00:00:00')

In [80]:
df_sessions_synth['SessionDate'].min()

Timestamp('2023-09-01 00:00:00')

In [81]:
df_transactions_synth = df_transactions_synth[df_transactions_synth['TransactionDate'] > '2024-08-31']
df_sessions_synth = df_sessions_synth[df_sessions_synth['SessionDate'] > '2024-08-31']

In [82]:
df_clients.to_parquet(DATA_PATH/'clients.parquet', index=False)
df_transactions_synth.to_parquet(DATA_PATH/'transactions.parquet', index=False)
df_sessions_synth.to_parquet(DATA_PATH/'app_sessions.parquet', index=False)

# <font color="#6e8ebc">**Datasets Split**</font>

## <font color="#ff743d">**Transactions**</font>

In [83]:
df_transactions_synth.head(10)

,TransactionId,CustomerId,TransactionDate,Amount,TransactionType
1858355,1039759,15793307,2024-09-01,2031.92,TRANSFER
1299413,1201904,15723856,2024-09-01,325.08,TRANSFER
1222378,894832,15714391,2024-09-01,50.84,DEBIT
874978,1596419,15672499,2024-09-01,338.77,TRANSFER
1062380,780309,15694805,2024-09-01,784.63,TRANSFER
1220484,328079,15714080,2024-09-01,996.37,CASH_IN
888857,247521,15674148,2024-09-01,103.11,PAYMENT
1860931,1179531,15793478,2024-09-01,133.32,CASH_IN
793613,812259,15663164,2024-09-01,1228.15,TRANSFER
1684920,1954254,15772009,2024-09-01,113.56,DEBIT


In [84]:
df_transactions_synth

,TransactionId,CustomerId,TransactionDate,Amount,TransactionType
1858355,1039759,15793307,2024-09-01,2031.92,TRANSFER
1299413,1201904,15723856,2024-09-01,325.08,TRANSFER
1222378,894832,15714391,2024-09-01,50.84,DEBIT
874978,1596419,15672499,2024-09-01,338.77,TRANSFER
1062380,780309,15694805,2024-09-01,784.63,TRANSFER
...,...,...,...,...,...
938742,584402,15680346,2025-09-30,410.27,CASH_IN
1426394,877855,15740175,2025-09-30,124.89,PAYMENT
938741,584398,15680346,2025-09-30,81.68,PAYMENT
1421713,497231,15739502,2025-09-30,175.97,TRANSFER


In [85]:
tx_2024_Sep = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2024-09-01') & (df_transactions_synth['TransactionDate'] <= '2024-09-30')]
tx_2024_Oct = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2024-10-01') & (df_transactions_synth['TransactionDate'] <= '2024-10-31')]
tx_2024_Nov = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2024-11-01') & (df_transactions_synth['TransactionDate'] <= '2024-11-30')]
tx_2024_Dic = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2024-12-01') & (df_transactions_synth['TransactionDate'] <= '2024-12-31')]
tx_2025_Ene = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2025-01-01') & (df_transactions_synth['TransactionDate'] <= '2025-01-30')]
tx_2025_Feb = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2025-02-01') & (df_transactions_synth['TransactionDate'] <= '2025-02-28')]
tx_2025_Mar = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2025-03-01') & (df_transactions_synth['TransactionDate'] <= '2025-03-31')]
tx_2025_Abr = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2025-04-01') & (df_transactions_synth['TransactionDate'] <= '2025-04-30')]
tx_2025_May = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2025-05-01') & (df_transactions_synth['TransactionDate'] <= '2025-05-31')]
tx_2025_Jun = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2025-06-01') & (df_transactions_synth['TransactionDate'] <= '2025-06-30')]
tx_2025_Jul = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2025-07-01') & (df_transactions_synth['TransactionDate'] <= '2025-07-31')]
tx_2025_Ago = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2025-08-01') & (df_transactions_synth['TransactionDate'] <= '2025-08-31')]
tx_2025_Sep = df_transactions_synth[(df_transactions_synth['TransactionDate'] >= '2025-09-01') & (df_transactions_synth['TransactionDate'] <= '2025-09-30')]

In [86]:
tx_2024_Sep.to_parquet(TX_DATA_PATH / 'tx_2024_Sep.parquet', index=False)
tx_2024_Oct.to_parquet(TX_DATA_PATH / 'tx_2024_Oct.parquet', index=False)
tx_2024_Nov.to_parquet(TX_DATA_PATH / 'tx_2024_Nov.parquet', index=False)
tx_2024_Dic.to_parquet(TX_DATA_PATH / 'tx_2024_Dic.parquet', index=False)
tx_2025_Ene.to_parquet(TX_DATA_PATH / 'tx_2025_Ene.parquet', index=False)
tx_2025_Feb.to_parquet(TX_DATA_PATH / 'tx_2025_Feb.parquet', index=False)
tx_2025_Mar.to_parquet(TX_DATA_PATH / 'tx_2025_Mar.parquet', index=False)
tx_2025_Abr.to_parquet(TX_DATA_PATH / 'tx_2025_Abr.parquet', index=False)
tx_2025_May.to_parquet(TX_DATA_PATH / 'tx_2025_May.parquet', index=False)
tx_2025_Jun.to_parquet(TX_DATA_PATH / 'tx_2025_Jun.parquet', index=False)
tx_2025_Jul.to_parquet(TX_DATA_PATH / 'tx_2025_Jul.parquet', index=False)
tx_2025_Ago.to_parquet(TX_DATA_PATH / 'tx_2025_Ago.parquet', index=False)
tx_2025_Sep.to_parquet(TX_DATA_PATH / 'tx_2025_Sep.parquet', index=False)

## <font color="#ff743d">**App Interaction**</font>

In [87]:
df_sessions_synth['SessionDate'].max()

Timestamp('2025-09-30 00:00:00')

In [88]:
df_sessions_synth.head(10)

,SessionId,CustomerId,SessionDate,DurationMin,UsedTransfer,UsedPayment,UsedInvest,OpenedPush,FailedLogin,month
568575,568575,15743490,2024-09-01,47.973278,0,1,0,0,0,2024-09-01
700155,700155,15706269,2024-09-01,38.137272,0,0,0,1,0,2024-09-01
912306,912306,15794360,2024-09-01,28.483303,0,0,0,0,0,2024-09-01
186350,186350,15596414,2024-09-01,4.753233,0,0,0,0,0,2024-09-01
909970,909970,15799534,2024-09-01,5.700523,0,0,0,0,0,2024-09-01
142412,142412,15682773,2024-09-01,19.884031,0,0,0,0,0,2024-09-01
236732,236732,15583548,2024-09-01,18.125655,1,0,0,1,0,2024-09-01
1059210,1059210,15580462,2024-09-01,14.000372,0,0,1,0,0,2024-09-01
788149,788149,15568878,2024-09-01,73.876609,0,0,0,0,0,2024-09-01
689282,689282,15699335,2024-09-01,7.108149,0,0,0,1,0,2024-09-01


In [89]:
ss_2024_Sep = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2024-09-01') & (df_sessions_synth['SessionDate'] <= '2024-09-30')]
ss_2024_Oct = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2024-10-01') & (df_sessions_synth['SessionDate'] <= '2024-10-31')]
ss_2024_Nov = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2024-11-01') & (df_sessions_synth['SessionDate'] <= '2024-11-30')]
ss_2024_Dic = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2024-12-01') & (df_sessions_synth['SessionDate'] <= '2024-12-31')]
ss_2025_Ene = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2025-01-01') & (df_sessions_synth['SessionDate'] <= '2025-01-30')]
ss_2025_Feb = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2025-02-01') & (df_sessions_synth['SessionDate'] <= '2025-02-28')]
ss_2025_Mar = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2025-03-01') & (df_sessions_synth['SessionDate'] <= '2025-03-31')]
ss_2025_Abr = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2025-04-01') & (df_sessions_synth['SessionDate'] <= '2025-04-30')]
ss_2025_May = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2025-05-01') & (df_sessions_synth['SessionDate'] <= '2025-05-31')]
ss_2025_Jun = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2025-06-01') & (df_sessions_synth['SessionDate'] <= '2025-06-30')]
ss_2025_Jul = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2025-07-01') & (df_sessions_synth['SessionDate'] <= '2025-07-31')]
ss_2025_Ago = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2025-08-01') & (df_sessions_synth['SessionDate'] <= '2025-08-31')]
ss_2025_Sep = df_sessions_synth[(df_sessions_synth['SessionDate'] >= '2025-09-01') & (df_sessions_synth['SessionDate'] <= '2025-09-30')]

In [90]:
ss_2024_Sep.to_parquet(APP_DATA_PATH / 'ss_2024_Sep.parquet', index=False)
ss_2024_Oct.to_parquet(APP_DATA_PATH / 'ss_2024_Oct.parquet', index=False)
ss_2024_Nov.to_parquet(APP_DATA_PATH / 'ss_2024_Nov.parquet', index=False)
ss_2024_Dic.to_parquet(APP_DATA_PATH / 'ss_2024_Dic.parquet', index=False)
ss_2025_Ene.to_parquet(APP_DATA_PATH / 'ss_2025_Ene.parquet', index=False)
ss_2025_Feb.to_parquet(APP_DATA_PATH / 'ss_2025_Feb.parquet', index=False)
ss_2025_Mar.to_parquet(APP_DATA_PATH / 'ss_2025_Mar.parquet', index=False)
ss_2025_Abr.to_parquet(APP_DATA_PATH / 'ss_2025_Abr.parquet', index=False)
ss_2025_May.to_parquet(APP_DATA_PATH / 'ss_2025_May.parquet', index=False)
ss_2025_Jun.to_parquet(APP_DATA_PATH / 'ss_2025_Jun.parquet', index=False)
ss_2025_Jul.to_parquet(APP_DATA_PATH / 'ss_2025_Jul.parquet', index=False)
ss_2025_Ago.to_parquet(APP_DATA_PATH / 'ss_2025_Ago.parquet', index=False)
ss_2025_Sep.to_parquet(APP_DATA_PATH / 'ss_2025_Sep.parquet', index=False)